File: packages/storage/amplify_storage_s3_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_storage_s3/types/FlutterDownloadFileRequest.kt
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_storage_s3.types

import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
import com.amplifyframework.storage.StorageAccessLevel
import com.amplifyframework.storage.options.StorageDownloadFileOptions
import java.io.File
import java.util.*

data class FlutterDownloadFileRequest(val request: Map<String, *>) {
    val uuid: String = request["uuid"] as String
    val key: String = request["key"] as String
    val file: File = File(request["path"] as String)
    val options: StorageDownloadFileOptions = setOptions(request)

    private fun setOptions(request: Map<String, *>): StorageDownloadFileOptions {
        if (request["options"] != null) {
            val optionsMap = request["options"] as Map<String, *>
            val options: StorageDownloadFileOptions.Builder<*> = StorageDownloadFileOptions.builder()

            optionsMap.forEach { (optionKey, optionValue) ->
                when (optionKey) {
                    "accessLevel" -> {
                        val accessLevelStringOption = optionValue as String
                        val accessLevel: StorageAccessLevel? = StorageAccessLevel.values().find {
                            it.name == accessLevelStringOption.uppercase(
                                Locale.ROOT
                            )
                        }
                        options.accessLevel(accessLevel)
                    }
                    "targetIdentityId" -> {
                        options.targetIdentityId(optionValue as String?)
                    }
                }
            }
            return options.build()
        }
        return StorageDownloadFileOptions.defaultInstance()
    }

    companion object {
        private const val validationErrorMessage: String = "DownloadFile request malformed."
        fun validate(request: Map<String, *>) {
            if (request["uuid"] !is String) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("uuid"))
            }
            if (request["path"] !is String) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("path"))
            }
            if (request["key"] !is String) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("key"))
            }
        }
    }
}

File: packages/storage/amplify_storage_s3_ios/ios/Classes/FlutterDownloadFileRequest.swift
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify
import amplify_flutter_ios

struct FlutterDownloadFileRequest {
    var uuid: String
    var key: String
    var file: URL
    var options: StorageDownloadFileRequest.Options?
    init(request: Dictionary<String, AnyObject>) {
        self.uuid = request["uuid"] as! String
        self.key = request["key"] as! String
        self.file = NSURL(fileURLWithPath: request["path"] as! String) as URL
        self.options = setOptions(request: request)
    }
    
    static func validate(request: Dictionary<String, AnyObject>) throws {
        let validationErrorMessage = "DownloadFile request malformed."
        if !(request["uuid"] is String) {
            throw InvalidRequestError.storage(comment: validationErrorMessage,
                                              suggestion: String(format: ErrorMessages.missingAttribute, "uuid"))
        }
        if !(request["key"] is String) {
            throw InvalidRequestError.storage(comment: validationErrorMessage,
                                              suggestion: String(format: ErrorMessages.missingAttribute, "key"))
        }
        if !(request["path"] is String) {
            throw InvalidRequestError.storage(comment: validationErrorMessage,
                                              suggestion: String(format: ErrorMessages.missingAttribute, "path"))
        }
    }
    
    private func setOptions(request: Dictionary<String, AnyObject>) -> StorageDownloadFileRequest.Options? {
        
        if(request["options"] != nil) {
            let requestOptions = request["options"] as! Dictionary<String, AnyObject>
            //Default options
            var accessLevel = StorageAccessLevel.guest
            var targetIdentityId: String? = nil
            
            for(key,value) in requestOptions {
                switch key {
                case "accessLevel":
                    accessLevel = StorageAccessLevel(rawValue: value as! String) ?? accessLevel
                case "targetIdentityId":
                    targetIdentityId = value as? String
                default:
                    print("Received unexpected option: \(key)")
                }
            }
            return StorageDownloadFileRequest.Options(accessLevel: accessLevel, targetIdentityId: targetIdentityId)
        }
        return nil
    }
}
